### ISSUE TITLE
Status of virtual bootstraps?

### ISSUE THREAD

**Issue by clauswilke** [DavisVaughan/strapgod#1]
I was wondering what the status of virtual bootstraps is. The C++ check that prevented this code from running in the first place seems to have been removed in the currently released dplyr (0.7.8), but something else has changed and virtual groups still don't work. Are there plans to revisit this or is this currently not on the roadmap? I'm asking because I'd like to figure out how much I should push my own bootstrap solution forward.

``` r
library(strapgod)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

iris %>%
  bootstrapify(10)
#> # A tibble: 150 x 5
#> # Groups:   [?]
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # ... with 140 more rows

iris %>%
  bootstrapify(10) %>%
  summarise(per_strap_mean = mean(Petal.Width))
#> # A tibble: 1 x 1
#> # Groups:   [?]
#>   per_strap_mean
#>            <dbl>
#> 1           1.20

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  summarise(per_strap_species_mean = mean(Petal.Width))
#> Error: 'group_data' is not an exported object from 'namespace:dplyr'
```

<sup>Created on 2018-12-19 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by DavisVaughan** [comment]
`group_data()` is coming in 0.8! (CMD+F for it in the news here https://github.com/tidyverse/dplyr/blob/master/NEWS.md)

I would like to make this a part of rsample when dplyr 0.8 comes out. I'm not sure exactly when that will be, but Romain ran revdeps on it in the past day or two so hopefully not too long from now (maybe before rstudio conf?)

**Comment by clauswilke** [comment]
That's great! Would you mind setting the remotes on this repo so it pulls in the current dplyr code then?

I plan to talk about bootstrapping at rstudio conf, as part of making hypothetical outcomes plots. I'd rather talk about this planned future feature, even if it isn't properly integrated/released yet, than about my current approach which is going to be outdated soon. I'm also thinking about reimplementing my functions `bootstrap_summarize()` and `bootstrap_do()` using this functionality. In this way, I can point people to my ungeviz package (which I expect will continue to be around in some form for a while) rather than this package which I assume will go away once the functionality is integrated into rsample.

**Comment by DavisVaughan** [comment]
Done! All readme examples seem to be working again

``` r
library(strapgod)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

iris %>%
  bootstrapify(10)
#> # A tibble: 150 x 5
#> # Groups:   .bootstrap [10]
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 140 more rows

iris %>%
  bootstrapify(10) %>%
  summarise(per_strap_mean = mean(Petal.Width))
#> # A tibble: 10 x 2
#>    .bootstrap per_strap_mean
#>         <int>          <dbl>
#>  1          1           1.22
#>  2          2           1.10
#>  3          3           1.35
#>  4          4           1.29
#>  5          5           1.20
#>  6          6           1.24
#>  7          7           1.24
#>  8          8           1.17
#>  9          9           1.13
#> 10         10           1.20

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  summarise(per_strap_species_mean = mean(Petal.Width))
#> # A tibble: 30 x 3
#> # Groups:   Species [3]
#>    Species .bootstrap per_strap_species_mean
#>    <fct>        <int>                  <dbl>
#>  1 setosa           1                  0.254
#>  2 setosa           2                  0.228
#>  3 setosa           3                  0.258
#>  4 setosa           4                  0.236
#>  5 setosa           5                  0.248
#>  6 setosa           6                  0.218
#>  7 setosa           7                  0.25 
#>  8 setosa           8                  0.228
#>  9 setosa           9                  0.262
#> 10 setosa          10                  0.262
#> # … with 20 more rows

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  collect()
#> # A tibble: 1,500 x 6
#> # Groups:   Species, .bootstrap [30]
#>    .bootstrap Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>         <int>        <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          1          5.1         3.8          1.5         0.3 setosa 
#>  2          1          4.9         3.1          1.5         0.1 setosa 
#>  3          1          5.8         4            1.2         0.2 setosa 
#>  4          1          4.8         3.4          1.6         0.2 setosa 
#>  5          1          4.9         3.1          1.5         0.2 setosa 
#>  6          1          5.7         3.8          1.7         0.3 setosa 
#>  7          1          5.2         4.1          1.5         0.1 setosa 
#>  8          1          4.9         3.1          1.5         0.2 setosa 
#>  9          1          5.2         4.1          1.5         0.1 setosa 
#> 10          1          4.7         3.2          1.6         0.2 setosa 
#> # … with 1,490 more rows

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  collect(.id = ".id", .original_id = ".original_id")
#> # A tibble: 1,500 x 8
#> # Groups:   Species, .bootstrap [30]
#>    .bootstrap   .id .original_id Sepal.Length Sepal.Width Petal.Length
#>         <int> <int>        <int>        <dbl>       <dbl>        <dbl>
#>  1          1     1           31          4.8         3.1          1.6
#>  2          1     2           11          5.4         3.7          1.5
#>  3          1     3           45          5.1         3.8          1.9
#>  4          1     4           50          5           3.3          1.4
#>  5          1     5            3          4.7         3.2          1.3
#>  6          1     6           24          5.1         3.3          1.7
#>  7          1     7            2          4.9         3            1.4
#>  8          1     8           45          5.1         3.8          1.9
#>  9          1     9           39          4.4         3            1.3
#> 10          1    10           30          4.7         3.2          1.6
#> # … with 1,490 more rows, and 2 more variables: Petal.Width <dbl>,
#> #   Species <fct>
```

<sup>Created on 2018-12-20 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by clauswilke** [comment]
> Go for it, this package will definitely disappear when it gets merged into rsample. Copy and paste the guts of this package to your heart's content.

Actually, I was just planning to call the `bootstrapify()` function from this package. This gives me a clean implementation and I just need to change the dependency once it's in rsample. Do you expect that `.n` and `.bootstrap` will be the final names for the parameter indicating the number of replicates and the data column indicating the boostrap id, respectively? Are these consistent with names in rsample?

**Comment by DavisVaughan** [comment]
That makes sense. I've updated the argument names to match `rsample` (removed `.` prefix) and to match `dplyr::collect()`. These are much more likely to stay the same, but I can't make any promises. 

There are a few other neat things you can do with the new `group_*()` functions in dplyr 0.8, but there are a few bugs on the dplyr end that I have reported, so its a bit unstable (see some of the new features on the updated readme in this repo)

I'll leave this open so I can remember to let you know when things move over to rsample

**Comment by clauswilke** [comment]
Thanks for all your efforts on this. The `collect()` functionality with `id` and `original_id` is very useful.

One minor comment: I think `rsample` uses `times` instead of `n`. https://github.com/tidymodels/rsample/blob/60fa7ec44c39c37907d7aad3559f53c45ad6e90e/R/boot.R#L11

**Comment by clauswilke** [comment]
Looks good! I ported my convenience functions over, and it works fine:

https://github.com/clauswilke/ungeviz/blob/736313554139268c0a394fed1927fdee979fc493/R/bootstrap-summarize.R#L27-L35

One question about `group_map()`: Isn't that basically `do()`? I've always felt that `do()` is an underappreciated member of the tidyverse.

``` r
library(dplyr)
library(broom)
library(strapgod)

set.seed(1234)

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  group_map(~tidy(lm(Sepal.Length ~ Petal.Length, data = .x)))
#> # A tibble: 60 x 7
#> # Groups:   Species, .bootstrap [30]
#>    Species .bootstrap term         estimate std.error statistic  p.value
#>  * <fct>        <int> <chr>           <dbl>     <dbl>     <dbl>    <dbl>
#>  1 setosa           1 (Intercept)    4.52       0.499     9.06  5.87e-12
#>  2 setosa           1 Petal.Length   0.397      0.344     1.15  2.54e- 1
#>  3 setosa           2 (Intercept)    4.38       0.470     9.32  2.42e-12
#>  4 setosa           2 Petal.Length   0.428      0.312     1.37  1.76e- 1
#>  5 setosa           3 (Intercept)    5.14       0.442    11.6   1.47e-15
#>  6 setosa           3 Petal.Length  -0.0877     0.302    -0.291 7.73e- 1
#>  7 setosa           4 (Intercept)    4.13       0.505     8.18  1.17e-10
#>  8 setosa           4 Petal.Length   0.611      0.344     1.78  8.19e- 2
#>  9 setosa           5 (Intercept)    5.19       0.427    12.2   2.92e-16
#> 10 setosa           5 Petal.Length  -0.0537     0.288    -0.187 8.53e- 1
#> # ... with 50 more rows

set.seed(1234)

iris %>%
  group_by(Species) %>%
  bootstrapify(10) %>%
  do(tidy(lm(Sepal.Length ~ Petal.Length, data = .)))
#> # A tibble: 60 x 7
#> # Groups:   Species, .bootstrap [30]
#>    Species .bootstrap term         estimate std.error statistic  p.value
#>    <fct>        <int> <chr>           <dbl>     <dbl>     <dbl>    <dbl>
#>  1 setosa           1 (Intercept)    4.52       0.499     9.06  5.87e-12
#>  2 setosa           1 Petal.Length   0.397      0.344     1.15  2.54e- 1
#>  3 setosa           2 (Intercept)    4.38       0.470     9.32  2.42e-12
#>  4 setosa           2 Petal.Length   0.428      0.312     1.37  1.76e- 1
#>  5 setosa           3 (Intercept)    5.14       0.442    11.6   1.47e-15
#>  6 setosa           3 Petal.Length  -0.0877     0.302    -0.291 7.73e- 1
#>  7 setosa           4 (Intercept)    4.13       0.505     8.18  1.17e-10
#>  8 setosa           4 Petal.Length   0.611      0.344     1.78  8.19e- 2
#>  9 setosa           5 (Intercept)    5.19       0.427    12.2   2.92e-16
#> 10 setosa           5 Petal.Length  -0.0537     0.288    -0.187 8.53e- 1
#> # ... with 50 more rows
```

<sup>Created on 2018-12-21 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by DavisVaughan** [comment]
I think it is becoming the "blessed" version of `do()` alongside `group_split()` and `group_nest()` and `group_walk()`. It has more restrictive input/output types, and gives you access to the groups with `.y`. The docs on `group_map` are pretty useful, but I see the confusion.

See also
https://github.com/tidyverse/dplyr/blob/master/R/do.r#L20

**Comment by clauswilke** [comment]
One other issue I noticed: The `bootstrapify()` function enquos `key`, but the `collect()` function doesn't seem to enquo `id` or `original_id`. This gets confusing when one wants to write a function that calls both of these and accepts all these arguments. I'm not sure what the right approach is here, but as is it seems strange. Since `bootstrapify()` is new, maybe not enquoing `key` and instead treating it like a regular variable holding a string would be better?

**Comment by clauswilke** [comment]
Alternatively, could `collect()` be changed?

**Comment by clauswilke** [comment]
To give you some context, this is the function where I ran into this issue:
https://github.com/clauswilke/ungeviz/blob/master/R/bootstrapper.R
All the various special columns (.row, .id, .bootstrap, etc) are useful when making animations with gganimate.

**Comment by clauswilke** [comment]
Yes. I came to the same conclusion in my code:
https://github.com/clauswilke/ungeviz/blob/7e5bc6a336ff7a03d3904a2d630ecb4736c4c0b4/R/bootstrapper.R#L78-L118

Since `key` specifies a label and not an expression that needs to be evaluated, I think taking a simple string argument (without `enquo()`) is the right way to go.

**Comment by DavisVaughan** [comment]
strapgod just hit cran. we decided to leave it in a separate package than rsample because the api is so different. it also lets strapgod have the freedom to be a bit more experimental and break things if necessary https://cran.r-project.org/web/packages/strapgod/

not much changed, but the internals were cleaned up and simplified a lot. there was a slight change in the arguments (`...` were added to force naming of arguments). this shouldn't break your code unless you weren't naming the "details" related arguments already. but just a heads up.
